helix-term/src/application.rs

```diff
@@ -807,25 +810,39 @@ impl Application {
                                         tags,
                                         source: diagnostic.source.clone(),
                                         data: diagnostic.data.clone(),
+                                        language_server_id: server_id,
                                     })
                                 })
                                 .collect();
 
-                            doc.set_diagnostics(diagnostics);
+                            doc.replace_diagnostics(diagnostics, server_id);
                         }
 
                         // Sort diagnostics first by severity and then by line numbers.
                         // Note: The `lsp::DiagnosticSeverity` enum is already defined in decreasing order
                         params
                             .diagnostics
                             .sort_unstable_by_key(|d| (d.severity, d.range.start));
+                        let diagnostics = params
+                            .diagnostics
+                            .into_iter()
+                            .map(|d| (d, server_id, offset_encoding))
+                            .collect();
 
                         // Insert the original lsp::Diagnostics here because we may have no open document
                         // for diagnosic message and so we can't calculate the exact position.
                         // When using them later in the diagnostics picker, we calculate them on-demand.
-                        self.editor
-                            .diagnostics
-                            .insert(params.uri, params.diagnostics);
+                        match self.editor.diagnostics.entry(params.uri) {
+                            Entry::Occupied(o) => {
+                                let current_diagnostics = o.into_mut();
+                                // there may entries of other language servers, which is why we can't overwrite the whole entry
+                                current_diagnostics.retain(|(_, lsp_id, _)| *lsp_id != server_id);
+                                current_diagnostics.extend(diagnostics);
```

Comment:
This approach works for now but I wonder if this should become `HashMap<namespace, Vec<Diagnostic>>` to avoid a single array.

Also, I think this breaks the assumption in a couple places that diagnostics are sorted? Since we use `binary_search` on it

Reply:
Ah I see, there's `replace_diagnostics` but it's not used here. I think a single Vec<> still makes sense so we can binary search across the range

Reply:
Ah, I've just seen, that the diagnostics are sorted just above it (how could I've missed that...)
Yeah I agree to just use a `Vec` for now, I think I thought about, or even tried to use a HashMap (I'm not sure anymore it was some time ago), but a `Vec` seemed simpler and more appropriate that time.